module/zfs/vdev.c

```diff
@@ -898,6 +898,14 @@ vdev_alloc(spa_t *spa, vdev_t **vdp, nvlist_t *nv, vdev_t *parent, uint_t id,
 	(void) nvlist_lookup_uint64(nv, ZPOOL_CONFIG_CREATE_TXG,
 	    &vd->vdev_crtxg);
 
+	if (vd->vdev_ops == &vdev_root_ops &&
+	    (alloctype == VDEV_ALLOC_LOAD ||
+	    alloctype == VDEV_ALLOC_SPLIT ||
+	    alloctype == VDEV_ALLOC_ROOTPOOL)) {
+		(void) nvlist_lookup_uint64(nv,
+		    ZPOOL_CONFIG_VDEV_ROOT_ZAP, &spa->spa_root_vdev_zap);
+	}
```

Comment:
```suggestion
	} else {
		ASSERT0(spa->spa_root_vdev_zap);
	}
```

I'd suggest adding the `ASSERT0` here just to be consistent with the top-level and leaf zap handling.